### DISCUSSION TITLE
How to deal with relationships normally modelled with junction tables and joins?

### DISCUSSION THREAD

**Discussion by ollema** [pocketbase/pocketbase#289] [Q&A] [answered]
Hi!

I will try my best to describe my problem here - but I have a feeling that it might get confusing.

Consider that you have two collections/tables:
```
+-----------------+
| Events          |
+--------+--------+
| id     | string |
| name   | string |
+--------+--------+
```
```
+-----------------+
| Users           |
+--------+--------+
| id     | string |
| name   | string |
+--------+--------+
```

**Problem statement:**
I want every user to be able to respond with either `'interested'` or `'going'` to each event.

I also want to be able to query for a list of events with all user responses to be able to display how many and which users are going to each event in the event list view.

**How it could be solved "in SQL":**
Normally I would solve this with a junction table `responses`:

```sql
CREATE TABLE IF NOT EXISTS events (
    event_id BIGSERIAL PRIMARY KEY,
    event_name TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS users (
    user_id BIGSERIAL PRIMARY KEY,
    user_name TEXT NOT NULL
);

CREATE TYPE response AS enum (
    'interested',
    'going'
);

CREATE TABLE IF NOT EXISTS responses (
    response_id bigint NOT NULL REFERENCES events (event_id) ON DELETE CASCADE,
    user_id bigint NOT NULL REFERENCES users (user_id) ON DELETE CASCADE,
    response response NOT NULL,
);
```
then it is pretty simple to query:
```sql
SELECT
    events.event_id,
    events.event_name,
    ARRAY_AGG(responses.response) AS resps,
    ARRAY_AGG(users.user_id) AS user_ids,
    ARRAY_AGG(users.user_name) AS event_user_names,
FROM events
    INNER JOIN responses ON events.event_id = responses.event_id
    INNER JOIN users ON event_user.user_id = users.user_id
GROUP BY
    events.event_id,
    events.event_name   
```
(with reservation for typos).

With this it is also possible to easily query for all events that a certain user is attending by doing a similar join over the junction table "in the other direction".

---

**How can I achieve something similar in PocketBase?**

I have tried (and tested) some approaches but nothing feels great so far.

**Approach 1**
```
+----------------------------------+
| Events                           |
+-------------+--------------------+
| id          | string             |
| name        | string             |
| responses   | relation(Respones) |
+-------------+--------------------+
```
```
+----------------------------+
| Responses                  |
+----------+-----------------+
| id       | string          |
| response | multiple choice |
| user     | relation(Users) |
+----------+-----------------+
```
```
+-----------------+
| Users           |
+--------+--------+
| id     | string |
| name   | string |
+--------+--------+
```
**pros:**
* possible to make a single query to get a list over events and responses by simply expanding over `responses` and `responses.user` (actually `responses.profile` but you know the drill)

**cons:**
* each new user response require first a creation of a response record in the Responses collection and then an update to the responses field in the events collection
* each response has no connection at all to the actual event, which means...
* ... it's possible to end up with "orphan" responses if an event is deleted but the response is not. or if the creation of a response errors out after only the first half (see point 1).

---

**Approach 2**
```
+----------------------------------+
| Events                           |
+-------------+--------------------+
| id          | string             |
| name        | string             |
| responses   | relation(Respones) |
+-------------+--------------------+
```
```
+-----------------------------+
| Responses                   |
+----------+------------------+
| id       | string           |
| response | multiple choice  |
| event    | relation(Events) |
| user     | relation(Users)  |
+----------+------------------+
```
```
+-----------------+
| Users           |
+--------+--------+
| id     | string |
| name   | string |
+--------+--------+
```
**pros:**
* possible to make a single query to get a list over events and responses by simply expanding over `responses` and `responses.user` (actually `responses.profile` but you know the drill)
* now it's easier to prevent orphans by adding a cascade delete rule

**cons:**
* but this feels very wrong. is it very wrong? :smile: 
* each new response still requires two operations

---

**Approach 3**
```
+------------------------------------------+
| Events                                   |
+------------------------+-----------------+
| id                     | string          |
| name                   | string          |
| interested_responses   | relation(Users) |
| going_responses        | relation(Users) |
+------------------------+-----------------+
```
```
+-----------------+
| Users           |
+--------+--------+
| id     | string |
| name   | string |
+--------+--------+
```
**pros:**
* possible to make a single query to get a list over events and responses by simply expanding over `interested_responses` and `going_responses`
* no need to worry about orphans
* less collections, simpler logic in most cases...

**cons:**
* ...but have to make sure that each user only is present in one of these relations which makes it messy to add/change responses (since PocketBase does not support these kind of constraints)
* adding another response (e.g. `not going`) will require another field

---

A lot of text, I'm sorry. If you did make it through, thanks in advance and I'm curious to hear your thoughts. I've really enjoyed PocketBase so far and hopefully I can continue to use this for my project in some way :slightly_smiling_face:

**Comment by ganigeorgiev** [comment, marked as answer]
I personally would go with something like in "Approach 2", but without the `responses` relation in the events table (yes, you will have to make additional request(s) to get the `responses`, but maintaining the relation in 2 places is not ideal either).

Since this was asked previously, to improve a little the situation, I think we can implement the following:

1. Add support for virtual/computed/RawSql/etc. columns (https://github.com/pocketbase/pocketbase/discussions/265) but this may get tricky with how the API Rules will be applied but it will be useful also for creating a simple reporting pages when combined with https://github.com/pocketbase/pocketbase/issues/159.

2. Add support for **indirect expand** expressions, eg. `?indirectExpand=COLLECTION_NAME_OR_ID.REFERENCE_FIELD[.*].FINALIZER_PROP` that will get resolved similarly to the `expand` (if the indirect relation `viewRule` allows it).
  In the resolved JSON response we could add another reserved property `@indirectExpand` (or similar) to avoid collisions with the other fields available in the `@expand` object.

After v0.4.0 release, I'll investigate the above suggestions a little more in depth and may consider them for v0.6.0.

**Reply by ollema** [reply]
Thanks for your input - I will probably go with your suggestion.

A problem with this approach is that it's not only a single additional request when you list events + responses, each an additional request for every single event in that list.

In other words, the current list view for events (which is paginated with 20 records) makes 1 query to the database.

With the suggested approach this would be 21 queries.

Maybe this is not a big deal in the end, sqlite on the same server is pretty fast, but it does feel a bit wasteful.

---

As for 1) and 2), not sure if I fully understand how they would improve the situation described in this discussion :thinking: 

1) Sounds like it would be useful if I only wanted to display the number of responses but I will need to access user profiles for each response to each event as well (to display avatars etc)

2) This sounds super cool if I understood it correctly, would this be equivalent to the inner joins I used in the sql example above? Or what exactly do you mean with indirect expansion?

**Reply by ganigeorgiev** [reply]
> With the suggested approach this would be 21 queries.

You can make individual requests for each event, but you can also eager load the responses in a single request manually by extracting first the event ids and then applying them as filter with an OR (eg. `||`) condition:
```
/api/collections/responses/records?filter=(event="ID1" || event="ID2" || event="ID3" || ...)`
```
Once fetched you'll have to group them manually (you could store them in a map indexed by the event id).

About the 2 suggestions:

1) Could allow to handle simple aggregations, eg. sum, count or to create a computed columns, but yes, for your use case this may not be very useful if you want to access the relation data in a single request.

2) Could resolve exactly the issue with manually fetching and grouping the relations. For your use case this could look something like this (_the format is just an example_):
```
/api/collections/events/records?indirectExpand=responses.event
```
Internally PocketBase will handle the join operation based on the `responses.event` field and will return something like:

```js
{
    "id": "abcd",
    "name": "My event name",
    "@indirectExpand": {
        "responses": [
            {
                "id": "response_id",
                "event": "abcd",
                "user": "user_id"
            }
        ]
    }
}
```
Or in other words - the goal is to have similar behavior like `expand`, just with additional 2 extra prefixed keys that points to the related collection and field (an additional finalizer prop key may also be needed in case of duplicated indirect expand fields from different collections but I'll think more on it once I start working on the feature).

**Reply by ollema** [reply]
Ah, very true! I might do just that!

---

Both 1) and 2) looks like amazing additions, crossing my fingers that you will have time to work on these in the future.

Thanks again!

**Reply by ollema** [reply]
@ganigeorgiev would it make sense to create issues for 1) and 2) to track them (either to put them in the roadmap or in the backlog?).

if so, I could help out with creating the issues unless you prefer to do that yourself

**Reply by ganigeorgiev** [reply]
I didn't create initially the issues because I'm not sure about the implementation details yet, but here are links https://github.com/pocketbase/pocketbase/issues/311 and https://github.com/pocketbase/pocketbase/issues/312 so that we can track them (_the requirements in the description are not fixed and may change depending how they will be implemented_).
